src/index.ts

```diff
@@ -1,26 +1,71 @@
+interface DetailedEncryptionResult {
+  vault: string;
+  extractedKeyString: string;
+}
+
 interface EncryptionResult {
   data: string;
   iv: string;
   salt?: string;
 }
 
+interface DetailedDecryptResult {
+  extractedKeyString: string;
+  vault: unknown;
+  salt: string;
+}
+
+const EXPORT_FORMAT = 'jwk';
+const DERIVED_KEY_FORMAT = 'AES-GCM';
+const STRING_ENCODING = 'utf-8';
+
 /**
  * Encrypts a data object that can be any serializable value using
  * a provided password.
  *
  * @param {string} password - password to use for encryption
  * @param {R} dataObj - data to encrypt
+ * @param {CryptoKey} key - a CryptoKey instance
+ * @param {string} salt - salt used to encrypt
  * @returns {Promise<string>} cypher text
  */
-async function encrypt<R>(password: string, dataObj: R): Promise<string> {
-  const salt = generateSalt();
+async function encrypt<R>(
+  password: string,
+  dataObj: R,
+  key: CryptoKey,
```

Comment:
Should this parameter be optional?

```suggestion
  key?: CryptoKey,
```